README.md
# btpd-el

This is an Emacs frontend for the
[Btpd BitTorrent client](https://github.com/btpd/btpd)
providing more or less intuitive and simple interface
to it's functions. It also communicates with
[Emacs-w3m web browser](https://www.emacswiki.org/emacs/emacs-w3m)
allowing to start torrent download just by clicking on a respective
 link. Torrent's content can be previewed as a virtual directory tree.

See [Debian package README](debian/README.Debian) for more details.


debian/changelog
btpd-el (1.9xenial1) xenial; urgency=low

  * Per distribution build.
  * Bumped standards version to 3.9.7.
  * Added emacsen-compat file to prevent an error message during
    installation.
  * Minor spelling fix in README.Debian suggested by lintian.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Thu, 22 Feb 2018 19:59:15 +0300

btpd-el (1.9) unstable; urgency=low

  * Torrent state display enhancement.
  * Estimated download time display.
  * Separate sections for leeching and seeding torrents.
  * Added summary traffic info.
  * Added hot key for buffer refresh.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Tue, 03 Oct 2017 23:34:16 +0300

btpd-el (1.8) unstable; urgency=low

  * Don't ignore referer when retrieving torrents.
  * Properly restore page view after torrent adding dialog.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Tue, 01 Sep 2015 13:03:00 +0300

btpd-el (1.7) unstable; urgency=low

  * Fixed torrent content extraction. Files in virtual directories are
    now alphabetically sorted.
  * Fixed file path parsing.
  * Better integration with Emacspeak.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 20 Jul 2014 10:14:30 +0400

btpd-el (1.6) unstable; urgency=low

  * Fixed treatment of compressed torrent files downloaded by clicking a
    link in w3m.
  * More consistent keymap definition for torrent view mode.
  * Main control panel revised. Active and inactive torrents are now
    placed in separate lists that can be individually shown or hidden.
  * Implemented navigation by sections in control panel using PGUP and
    PGDN keys.
  * Better integration with Emacspeak.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sat, 24 May 2014 09:05:16 +0400

btpd-el (1.5) unstable; urgency=low

  * The btpd-w3m module is left uncompiled to avoid dependency on the
    w3m package.
  * More appropriate dependencies.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Fri, 28 Jun 2013 15:22:57 +0400

btpd-el (1.4) unstable; urgency=low

  * More consistent keymap in the torrent view.
  * Implemented size calculation for selected items in the torrent view.
  * Fixed virtual directory structure construction.
  * After returning from a subdirectory in torrent view position is
    correctly restored.
  * Tracker URLs are included in torrent info.
  * Fixed interactive torrent adding.
  * Added keystroke to close a control panel.
  * More extensive description in README.Debian.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Wed, 12 Jun 2013 10:50:24 +0400

btpd-el (1.3) unstable; urgency=low

  * Implemented privileges acquisition via sudo when it is necessary for
    content deletion.
  * Removed confusing options.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sat, 08 Jun 2013 06:27:03 +0400

btpd-el (1.2) unstable; urgency=low

  * Fixed hot keys for dired.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Fri, 07 Jun 2013 06:44:52 +0400

btpd-el (1.1) unstable; urgency=low

  * Corrected communication speed units display.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Fri, 07 Jun 2013 06:17:12 +0400

btpd-el (1.0) unstable; urgency=low

  * Initial Release.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Thu, 06 Jun 2013 08:24:31 +0400


lisp/btpd-w3m.el
;;; btpd-w3m.el --- Btpd bindings for Emacs-w3m
;;; Author: Igor B. Poretsky <poretsky@mlbox.ru>
;;; Keywords: W3m, Btpd, BitTorrent client

;;{{{  Copyright

;;; Copyright (C) 2013  Igor B. Poretsky

;;; This file is free software; you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 2, or (at your option)
;;; any later version.
;;;
;;; This file is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with GNU Emacs; see the file COPYING.  If not, write to
;;; the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;;; Boston, MA 02111-1307, USA.

;;}}}
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;{{{ Introduction

;;; Commentary:

;;; This module is part of the Emacs frontend for the Btpd BitTorrent
;;; client. It provides seamless torrents previewing and downloading
;;; facilities for the Emacs-w3m web-browser.

;;; Code:

;;}}}
;;{{{ Requirements

(eval-when-compile
  (require 'cl))
(require 'custom)
(require 'w3m)
(require 'btpd-bindings)

;;}}}
;;{{{ Customizations

(defcustom btpd-w3m-auto-setup t
  "If not `nil' setup w3m automatically to pass torrents to btpd.
This is done by adding items to the end of `w3m-content-type-alist'
if there is no item for the respective content type already."
  :type 'boolean
  :group 'btpd)

;;}}}
;;{{{ Bittorrent content type viewer function

(defconst btpd-w3m-torrent-content-type "application/x-bittorrent"
  "BitTorrent content type.")

(defconst btpd-w3m-torrent-file-name-pattern "\\.torrent\\'"
  "Regexp matching BitTorrent file name.")

(defun btpd-w3m-add-url (url &rest ignore)
  "Add URL to btpd."
  (let ((w3m-current-buffer (current-buffer))
        (working-directory (make-temp-file "btpd-w3m-" t)))
    (set-file-modes working-directory
                    (file-modes-symbolic-to-number "go+rx" (file-modes working-directory)))
    (lexical-let ((file (expand-file-name "torrent" working-directory))
                  (pos (point-marker))
                  (url url)
                  (curl w3m-current-url))
      (w3m-process-with-null-handler
        (w3m-process-do-with-temp-buffer
            (success (progn
                       (w3m-clear-local-variables)
                       (setq w3m-current-url url)
                       (w3m-retrieve url t nil nil curl handler)))
          (if (not success)
              (delete-directory (file-name-directory file) t)
	    (let ((buffer-file-coding-system 'binary)
		  (coding-system-for-write 'binary)
		  jka-compr-compression-info-list
		  format-alist)
              (write-region (point-min) (point-max) file)
              (w3m-touch-file file (w3m-last-modified url)))
            (when (and (equal curl w3m-current-url)
                       (buffer-name (marker-buffer pos)))
              (with-current-buffer (marker-buffer pos)
                (goto-char pos)
                (w3m-refontify-anchor)))
            (when (zerop (call-process "gzip" nil nil nil "-t" file))
              (rename-file file (format "%s.gz" file))
              (call-process "gzip" nil nil nil "-d" (format "%s.gz" file)))
            (btpd-add file
                      (lambda ()
                        (when (and (boundp 'btpd-new-torrent)
                                   (vectorp btpd-new-torrent)
                                   (= (length btpd-new-torrent) 3)
                                   (stringp (aref btpd-new-torrent 1)))
                          (cd "~")
                          (delete-directory (file-name-directory (aref btpd-new-torrent 1)) t))
                        (switch-to-buffer (marker-buffer pos))))))))))

;;}}}
;;{{{ W3m setup for cooperation with btpd daemon

(defun btpd-w3m-setup ()
  "Setup emacs-w3m to pass BitTorrent URLs to Btpd."
  (when (and btpd-w3m-auto-setup
             (boundp 'w3m-content-type-alist)
             (not (assoc btpd-w3m-torrent-content-type w3m-content-type-alist)))
    (add-to-list 'w3m-content-type-alist
                 (list btpd-w3m-torrent-content-type
                       btpd-w3m-torrent-file-name-pattern
                       'btpd-w3m-add-url
                       nil)
                 'append)
    (add-to-list 'w3m-content-type-alist
                 (list "" "" nil nil)
                 'append)
    (custom-note-var-changed 'w3m-content-type-alist)))

(eval-after-load "w3m" '(btpd-w3m-setup))

;;}}}

(provide 'btpd-w3m)

;;; btpd-w3m.el ends here


lisp/btpd-info.el
;;; btpd-info.el --- Torrent info extraction utility
;;; Author: Igor B. Poretsky <poretsky@mlbox.ru>
;;; Keywords: Btpd, BitTorrent client

;;{{{  Copyright

;;; Copyright (C) 2013  Igor B. Poretsky

;;; This file is free software; you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 2, or (at your option)
;;; any later version.
;;;
;;; This file is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with GNU Emacs; see the file COPYING.  If not, write to
;;; the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;;; Boston, MA 02111-1307, USA.

;;}}}
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;{{{ Introduction

;;; Commentary:

;;; This module is part of the Emacs frontend for the Btpd BitTorrent
;;; client. It provides torrent info extraction facility. It utilizes
;;; btinfo executable for retrieving data from a torrent file.

;;; Code:

;;}}}
;;{{{ Requirements

(eval-when-compile
  (require 'cl))

;;}}}
;;{{{ Torrent info extraction

(defconst btpd-info-command "btinfo"
  "Torrent info retrieving command.")

(defconst btpd-info-name-extractor "^Name: \\(.*\\)$"
  "Regexp for torrent name extraction.")

(defconst btpd-info-size-extractor "^Total size: \\([0-9]+\\)$"
  "Regexp for total torrent size extraction.")

(defconst btpd-info-nfiles-extractor "^Number of files: \\([0-9]+\\)$"
  "Regexp for torrent content files number extraction.")

(defconst btpd-info-content-list-header "^Files:$"
  "Regexp matching file list header in the torrent info.")

(defconst btpd-info-file-info-extractor "^\\(.*\\) (\\([0-9]+\\))$"
  "Regexp for file name and size extraction from the torrent info.")

(defconst btpd-info-hash-extractor "^Info hash: \\(.*\\)$"
  "Regexp for hash value extractor.")

(defconst btpd-info-urls-detector "^Tracker URLs: \\[ \\(.*\\) ]$"
  "Regexp matching Tracker URLs info item.")

(defconst btpd-info-url-extractor "\\[ \\([^ ]+\\) ]"
  "Regexp for Tracker URL extraction.")

(defun btpd-info-fix-links (tree)
  "Fix links in the directory tree."
  (dolist (node tree)
    (when (and (listp (cdr node))
               (not (string-equal (car node) "..")))
      (let ((backlink (assoc ".." (cdr node))))
        (when backlink
          (setcdr backlink tree)))
      (btpd-info-fix-links (cdr node)))))

(defun btpd-info-extract (file)
  "Extract info from specified torrent file
and return it as a vector of 9 elements:

0. Torrent name.

1. Total size in bytes represented as string.

2. Total number of files represented as string.

3. File list.

4. Source torrent file attributes.

5. Path to the source torrent file.

6. Reserved. It is initialized by `nil'.

7. Torrent hash.

8. Tracker URLs list.

File list is organized hierarchically. Directories are represented by
cons cells containing the directory name in car and it's content in cdr.
Files are represented by cons cells with the name in car and size in cdr.
File size is expressed in bytes and represented by string.
All these lists are maintained in the reverse order."
  (let ((content (make-vector 9 nil)))
    (with-temp-buffer
      (unless (zerop (call-process btpd-info-command nil t nil file))
        (error (buffer-string)))
      (goto-char (point-min))
      (when (re-search-forward btpd-info-name-extractor nil t)
        (aset content 0 (match-string-no-properties 1)))
      (goto-char (point-min))
      (when (re-search-forward btpd-info-size-extractor nil t)
        (aset content 1 (match-string-no-properties 1)))
      (goto-char (point-min))
      (when (re-search-forward btpd-info-nfiles-extractor nil t)
        (aset content 2 (match-string-no-properties 1)))
      (goto-char (point-min))
      (when (re-search-forward btpd-info-hash-extractor nil t)
        (aset content 7 (match-string-no-properties 1)))
      (goto-char (point-min))
      (when (re-search-forward btpd-info-urls-detector nil t)
        (goto-char (match-beginning 1))
        (let ((bound (match-end 1))
              (urls nil))
          (while (re-search-forward btpd-info-url-extractor bound t)
            (add-to-list 'urls (match-string-no-properties 1) 'append))
          (aset content 8 urls)))
      (goto-char (point-min))
      (when (re-search-forward btpd-info-content-list-header nil t)
        (while (re-search-forward btpd-info-file-info-extractor nil t)
          (let ((path nil)
                (tree nil)
                (fn (match-string-no-properties 1)))
            (while fn
              (push (file-name-nondirectory (directory-file-name fn)) path)
              (setq fn (file-name-directory (directory-file-name fn))))
            (setq fn (cons nil (aref content 3)))
            (dolist (component path)
              (setq tree (cdr fn))
              (setcdr fn
                      (add-to-list 'tree (cons component nil) nil
                                   (lambda (x y)
                                     (string-equal (car x) (car y)))))
              (setq fn (assoc component tree))
              (unless (cdr fn)
                (setcdr fn (list (cons ".." tree)))))
            (setcdr fn (match-string-no-properties 2))
            (while (setq fn (assoc ".." tree))
              (setq tree (cdr fn)))
            (aset content 3 tree)))))
    (btpd-info-fix-links (aref content 3))
    (aset content 4 (file-attributes file 'string))
    (aset content 5 (expand-file-name file))
    content))

;;}}}

(provide 'btpd-info)

;;; btpd-info.el ends here


lisp/btpd-view.el
;;; btpd-view.el --- Torrent content viewer for Btpd
;;; Author: Igor B. Poretsky <poretsky@mlbox.ru>
;;; Keywords: Btpd, BitTorrent client

;;{{{  Copyright

;;; Copyright (C) 2013  Igor B. Poretsky

;;; This file is free software; you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 2, or (at your option)
;;; any later version.
;;;
;;; This file is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with GNU Emacs; see the file COPYING.  If not, write to
;;; the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;;; Boston, MA 02111-1307, USA.

;;}}}
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;{{{ Introduction

;;; Commentary:

;;; This module is part of the Emacs frontend for the Btpd
;;; BitTorrent client. It provides a convenient way to browse
;;; torrent contents just like a virtual directory tree.

;;; Code:

;;}}}
;;{{{ Requirements

(eval-when-compile
  (require 'cl))
(require 'dired)
(require 'dired-x)
(require 'btpd-utils)
(require 'btpd-bindings)

;;}}}
;;{{{ Displaying torrent content as virtual directory tree

(defconst btpd-view-time-format "%Y-%m-%d %R"
  "Time format for directory listings.")

(defvar btpd-view-torrent-info nil
  "Stores info about currently viewed torrent as a vector of 8 elements
initially provided by `btpd-info-extract' and then consulted
and modified by other functions. It's elements keep the following data:

0. Torrent name.

1. Total size in bytes represented as string.

2. Total number of files represented as string.

3. Hierarchical file list.

4. Source torrent file attributes with modification time in string format.

5. Displayed directory path (real or fake).

6. Content availability flag.

7. Torrent hash.

8. Tracker URLs list.")
(make-variable-buffer-local 'btpd-view-torrent-info)

(defvar btpd-view-keymap
  (let ((map (make-sparse-keymap)))
    (suppress-keymap map)
    (set-keymap-parent map dired-mode-map)
    (define-key map (kbd "z") 'btpd-view-size)
    (loop for op in
          '(dired-advertised-find-file
            dired-find-file
            dired-view-file)
          do
          (eval
           `(define-key map [remap ,op] 'btpd-view-visit-item)))
    (loop for op in
          '(dired-flag-file-deletion
            dired-do-flagged-delete
            dired-do-delete
            dired-do-chgrp
            dired-do-chmod
            dired-do-chown
            dired-do-rename
            dired-do-rename-regexp
            dired-do-touch
            dired-do-compress
            dired-do-byte-compile
            dired-do-query-replace-regexp
            dired-sort-toggle-or-edit
            wdired-change-to-wdired-mode)
          do
          (eval
           `(define-key map [remap ,op] 'btpd-view-disabled)))
    map)
  "Keymap for torrent content virtual tree navigation.")

(defun btpd-view-buffer-setup (torrent-info)
  "Being provided by the torrent info, fill current buffer with the
viewed torrent directory content and setup it for browsing.
Torrent data must conform to the format detailed
in the `btpd-view-torrent-info' description."
  (let ((inhibit-read-only t)
        (content nil)
        (attributes nil)
        (links-size 0)
        (uid-size 0)
        (gid-size 0)
        (flen-size 0))
    (dolist (node (aref torrent-info 3))
      (push (make-vector 7 nil) content)
      (setq attributes
            (or (and (aref torrent-info 6)
                     (file-attributes (expand-file-name (car node)
                                                        (aref torrent-info 5))
                                      'string))
                (aref torrent-info 4)))
      (if (not (stringp (nth 5 attributes)))
          (setcar (nthcdr 5 attributes) (format-time-string btpd-view-time-format (nth 5 attributes)))
        (setcar (cdr attributes)
                (let ((count 1))
                  (when (listp (cdr node))
                    (dolist (child (cdr node))
                      (when (and (listp (cdr child))
                                 (not (string-equal (car child) "..")))
                        (incf count))))
                  count))
        (setcar (nthcdr 8 attributes)
                (if (listp (cdr node))
                    "dr-xr-xr-x"
                  "----------")))
      (aset (car content) 0 (nth 8 attributes))
      (aset (car content) 1 (number-to-string (nth 1 attributes)))
      (when (> (length (aref (car content) 1)) links-size)
        (setq links-size (length (aref (car content) 1))))
      (aset (car content) 2
            (if (numberp (nth 2 attributes))
                (number-to-string (nth 2 attributes))
              (nth 2 attributes)))
      (when (> (length (aref (car content) 2)) uid-size)
        (setq uid-size (length (aref (car content) 2))))
      (aset (car content) 3
            (if (numberp (nth 3 attributes))
                (number-to-string (nth 3 attributes))
              (nth 3 attributes)))
      (when (> (length (aref (car content) 3)) gid-size)
        (setq gid-size (length (aref (car content) 3))))
      (aset (car content) 4
            (if (stringp (cdr node))
                (cdr node)
              "0"))
      (when (> (length (aref (car content) 4)) flen-size)
        (setq flen-size (length (aref (car content) 4))))
      (aset (car content) 5 (nth 5 attributes))
      (aset (car content) 6 (car node)))
    (erase-buffer)
    (when (fboundp 'remove-overlays)
      (remove-overlays))
    (insert "  " (aref torrent-info 5) ":\n")
    (dolist (url (aref torrent-info 8))
      (insert "  " url "\n"))
    (insert "  Total " (btpd-format-size (aref torrent-info 1))
            " in " (aref torrent-info 2) " file")
    (unless (= (string-to-number (aref torrent-info 2)) 1)
      (insert "s"))
    (insert "\n")
    (dolist (item content)
      (insert "  " (aref item 0)
              " "
              (format (concat "%" (number-to-string links-size) "s")
                      (aref item 1))
              " "
              (format (concat "%-" (number-to-string uid-size) "s")
                      (aref item 2))
              " "
              (format (concat "%-" (number-to-string gid-size) "s")
                      (aref item 3))
              " "
              (format (concat "%" (number-to-string flen-size) "s")
                      (aref item 4))
              " "
              (aref item 5)
              " "
              (aref item 6)
              "\n"))
    (set (make-local-variable 'dired-actual-switches) "-al")
    (dired-build-subdir-alist)
    (dired-virtual (aref torrent-info 5))
    (use-local-map btpd-view-keymap))
  (setq btpd-view-torrent-info torrent-info)
  (set (make-local-variable 'revert-buffer-function) 'btpd-view-revert))

;;}}}
;;{{{ Utility functions

(defun btpd-view-revert (&rest ignore)
  "The revert buffer function for the torrent view buffers."
  (let ((opoint (point))
        (ofile (dired-get-filename nil t))
        (mark-alist (dired-remember-marks (point-min) (point-max))))
    (btpd-view-buffer-setup btpd-view-torrent-info)
    (let ((inhibit-read-only t))
      (dired-mark-remembered mark-alist)
      (or (and ofile (dired-goto-file ofile))
          (goto-char opoint))
      (dired-move-to-filename))))

(defun btpd-view-get-size (nodes)
  "Calculate total size of the listed nodes.
Returns cons cell with total size in car and number of files in cdr."
  (let ((size 0)
        (files 0))
    (dolist (item nodes)
      (unless (string-equal (car item) "..")
        (if (listp (cdr item))
            (let ((augment (btpd-view-get-size (cdr item))))
              (incf size (car augment))
              (incf files (cdr augment)))
          (incf size (string-to-number (cdr item)))
          (incf files))))
    (cons size files)))

(defun btpd-view-validate-buffer ()
  "Ensure that we are in an appropriate buffer."
  (unless btpd-view-torrent-info
    (error "Not in a torrent view buffer")))

(defun btpd-view-get-filename ()
  "Get file name at point."
  (let ((filename (dired-get-filename 'verbatim t)))
    (unless filename
      (error "No file on this line"))
    filename))

(defun btpd-view-get-node ()
  "Get node at point or `nil' if none."
  (assoc (btpd-view-get-filename) (aref btpd-view-torrent-info 3)))

;;}}}
;;{{{ Interactive commands

(defun btpd-view (torrent-file &optional base-dir name)
  "View torrent content in virtual dired buffer.
Other arguments are optional. The second one provides
the base directory where actual content resides if it is available.
The third argument allows to specify an alternative torrent name
that will be inherited by the buffer displaying the content."
  (interactive "fTorrent file: ")
  (let ((torrent-info (btpd-info-extract (expand-file-name torrent-file))))
    (when base-dir
      (aset torrent-info 5 (expand-file-name base-dir))
      (aset torrent-info 6 t))
    (setcar (nthcdr 5 (aref torrent-info 4))
            (format-time-string btpd-view-time-format
                                (nth 5 (aref torrent-info 4))))
    (with-current-buffer (get-buffer-create (or name (aref torrent-info 0)))
      (btpd-view-buffer-setup torrent-info)
      (switch-to-buffer (current-buffer))
      (when (and (interactive-p)
                 (featurep 'emacspeak))
        (emacspeak-auditory-icon 'open-object)
        (emacspeak-speak-mode-line)))))

(defun btpd-view-visit-item ()
  "Visit file or directory at point."
  (interactive)
  (btpd-view-validate-buffer)
  (let ((node (btpd-view-get-node)))
    (if (and node (listp (cdr node)))
        (let ((position (and (string-equal (car node) "..")
                             (aref btpd-view-torrent-info 5))))
          (aset btpd-view-torrent-info 3 (cdr node))
          (aset btpd-view-torrent-info 5
                (expand-file-name (car node)
                                  (aref btpd-view-torrent-info 5)))
          (btpd-view-buffer-setup btpd-view-torrent-info)
          (when position
            (dired-goto-file position))
          (when (and (interactive-p)
                     (featurep 'emacspeak))
            (emacspeak-auditory-icon 'open-object)
            (emacspeak-speak-mode-line)))
      (if (interactive-p)
          (call-interactively 'dired-view-file)
        (dired-view-file)))))

(defun btpd-view-from-dired ()
  "Preview torrent content from a file in dired."
  (interactive)
  (let ((file (btpd-view-get-filename)))
    (if (file-regular-p file)
        (progn
          (btpd-view file)
          (when (and (interactive-p)
                     (featurep 'emacspeak))
            (emacspeak-auditory-icon 'open-object)
            (emacspeak-speak-mode-line)))
      (error "Not a regular file"))))

(defun btpd-view-size (&optional arg)
  "Calculate total size of all marked (or next ARG) files
or directories in a torrent view buffer. Parent directory link `..'
is always ignored even if it is selected."
  (interactive "P")
  (btpd-view-validate-buffer)
  (let ((emacspeak-speak-messages t)
        (amount (btpd-view-get-size (dired-map-over-marks (btpd-view-get-node) arg))))
    (message "%s in %d file%s"
             (btpd-format-size (car amount))
             (cdr amount)
             (if (= (cdr amount) 1)
                 ""
               "s"))))

(defun btpd-view-disabled ()
  "Signal a disabled operation."
  (interactive)
  (error "The operation is not available here"))

;;}}}

(provide 'btpd-view)

;;; btpd-view.el ends here


lisp/btpd-setup.el
;;; btpd-setup.el --- Emacs frontend for the Btpd BitTorrent client initial setup
;;; Author: Igor B. Poretsky <poretsky@mlbox.ru>
;;; Keywords: Btpd, BitTorrent client

;;{{{  Copyright

;;; Copyright (C) 2013  Igor B. Poretsky

;;; This file is free software; you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 2, or (at your option)
;;; any later version.
;;;
;;; This file is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with GNU Emacs; see the file COPYING.  If not, write to
;;; the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;;; Boston, MA 02111-1307, USA.

;;}}}
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;{{{ Introduction

;;; Commentary:

;;; This module contains all necessary autoloads and other initial actions
;;; that should be done to communicate comfortably with Btpd
;;; in Emacs. It should be required from an Emacs startup file.

;;; Code:

;;}}}
;;{{{ Requirements

(autoload 'btpd "btpd" "Btpd BitTorrent client" t)
(autoload 'btpd-add "btpd" "Add torrent file to btpd interactively." t)
(autoload 'btpd-add-from-dired "btpd" "Add torrent from a file in dired." t)
(autoload 'btpd-view "btpd-view" "View torrent content in virtual dired buffer." t)
(autoload 'btpd-view-from-dired "btpd-view" "Preview torrent content from a file in dired." t)

;;}}}
;;{{{ Setup for main menu

(define-key global-map [menu-bar btpd] '("Btpd" . btpd))

;;}}}
;;{{{ Setup for dired

(eval-after-load "dired"
  '(progn
     (define-key dired-mode-map (kbd "M-t") 'btpd-add-from-dired)
     (define-key dired-mode-map (kbd "M-RET") 'btpd-view-from-dired)))

;;}}}
;;{{{ Setup for w3m

(eval-after-load 'w3m '(require 'btpd-w3m nil t))

;;}}}

(provide 'btpd-setup)

;;; btpd-setup.el ends here


lisp/btpd-bindings.el
;;; btpd-bindings.el --- General bindings for btpd-el package
;;; Author: Igor B. Poretsky <poretsky@mlbox.ru>
;;; Keywords: Btpd, BitTorrent client

;;{{{  Copyright

;;; Copyright (C) 2013  Igor B. Poretsky

;;; This file is free software; you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 2, or (at your option)
;;; any later version.
;;;
;;; This file is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with GNU Emacs; see the file COPYING.  If not, write to
;;; the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;;; Boston, MA 02111-1307, USA.

;;}}}
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;{{{ Introduction

;;; Commentary:

;;; This module contains all necessary autoloads and other initial actions
;;; that should be done to communicate comfortably with Btpd
;;; in Emacs. It should be required from an Emacs startup file.

;;; Code:

;;}}}
;;{{{ Autoload interactive commands

(autoload 'btpd "btpd" "Btpd BitTorrent client" t)
(autoload 'btpd-add "btpd" "Add torrent file to btpd interactively." t)
(autoload 'btpd-add-from-dired "btpd" "Add torrent from a file in dired." t)
(autoload 'btpd-view "btpd-view" "View torrent content in virtual dired buffer." t)
(autoload 'btpd-view-from-dired "btpd-view" "Preview torrent content from a file in dired." t)

;;}}}
;;{{{ Add item to the main menu

(define-key global-map [menu-bar btpd] '("Btpd" . btpd))

;;}}}
;;{{{ Bindings for dired

(eval-after-load "dired"
  '(progn
     (define-key dired-mode-map (kbd "M-t") 'btpd-add-from-dired)
     (define-key dired-mode-map (kbd "M-RET") 'btpd-view-from-dired)))

;;}}}
;;{{{ Bindings for w3m

(eval-after-load 'w3m '(require 'btpd-w3m nil t))

;;}}}

(provide 'btpd-bindings)

;;; btpd-bindings.el ends here
